### ISSUE TITLE
[BUG] avoid-passing-async-when-sync-expected interprets FutureOr Functions as synchronous functions

### ISSUE THREAD

**Issue by NANASHI0X74** [dart-code-checker/dart-code-metrics#1033]
* **Dart code metrics version: 4.18.2**
* **Dart sdk version: 2.18.2**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>  

```yaml
dart_code_metrics:
  metrics:
    cyclomatic-complexity: 20
    number-of-parameters: 4
    maximum-nesting-level: 5
    source-lines-of-code: 35
  metrics-exclude:
    - test/**
    - config/**
    - lib/firebase_options.dart
  rules:
    - avoid-banned-imports:
        entries:
          - paths: [ "lib/.*" ]
            deny: [ "package:flutter/cupertino.dart" ]
            message: "Do not import Flutter Cupertino Design library, we should not depend on it! This is an Android only app"
    - avoid-nested-conditional-expressions:
        acceptable-level: 2
    - no-empty-block
    - prefer-conditional-expressions
    - avoid-global-state
    # - avoid-ignoring-return-values
    # - avoid-dynamic
    # - avoid-late-keyword
    # - member-ordering-extended
    # - no-magic-numbers
    # - prefer-correct-identifier-length
    # - prefer-correct-type-name
    # - prefer-enums-by-name
    # - avoid-returning-widgets
    # - avoid-shrink-wrap-in-lists
    - avoid-passing-async-when-sync-expected
    - avoid-missing-enum-constant-in-map
    - avoid-non-ascii-symbols
    - avoid-throw-in-catch-block
    - avoid-top-level-members-in-tests
    - avoid-unrelated-type-assertions
    - avoid-unused-parameters
    - no-equal-then-else
    - no-object-declaration
    - prefer-commenting-analyzer-ignores
    - prefer-conditional-expressions
    - prefer-first
    - prefer-last
    - prefer-immediate-return
    - prefer-match-file-name
    - prefer-moving-to-variable
    - always-remove-listener
    - avoid-unnecessary-setstate
    - avoid-use-expanded-as-spacer
    - avoid-wrapping-in-padding
    - prefer-correct-edge-insets-constructor
    - prefer-extracting-callbacks:
        allowed-line-count: 3
    - prefer-single-widget-per-file:
        ignore-private-widgets: true
  anti-patterns:
    - long-method
    - long-parameter-list

```

</details>

**What did you do? Please include the source code example causing the issue.**

Any asynchronous function passed to methods that take functions returning FutureOr - such as Future's then() and whenComplete() methods - will produce lint warnings, even though asynchronous functions are expected there.

as mentioned here: https://github.com/dart-code-checker/dart-code-metrics/pull/987#issuecomment-1276484353

This is just a minimal example:

```dart
Future<String> doSomeGetRequest() => Future.value('');

Future<String> doAnotherGetRequest(String input) => Future.value('');

Future<String> main() => doSomeGetRequest()
    .then(doAnotherGetRequest); // lint
```

**What did you expect to happen?**
no lint warning for passing async functions when async functions were expected

**What actually happened?**
lint warnings were produced

**Are you willing to submit a pull request to fix this bug?**
maybe- I'm unfamiliar with the codebase though

**Comment by incendial** [comment]
Fix available in 4.21.2 🚀
